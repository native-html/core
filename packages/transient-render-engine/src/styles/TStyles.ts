import { CSSProperties, CSSProcessedProps } from '@native-html/css-processor';
import nil from 'ramda/src/isNil';
import not from 'ramda/src/not';
import compose from 'ramda/src/compose';

const notNil = compose(not, nil);
/**
 * A merge properties from left to right.
 *
 * @param child
 * @param parent
 */
function inheritProperties(
  ...properties: Array<CSSProperties | null | undefined>
): CSSProperties {
  const realProperties = properties.filter(notNil) as CSSProperties[];
  if (realProperties.length === 1) {
    return realProperties[0];
  }
  if (Object.keys(realProperties[0]).length === 0) {
    return realProperties[1];
  }
  return realProperties.reduce(
    (prev, curr) => ({ ...prev, ...curr }),
    Object.prototype as CSSProperties
  ) as CSSProperties;
}

/**
 * A record of styles organized in logical chunks:
 *
 * - wether they are supported in React Native (native) and others (web).
 * - wether they target native Views (block) or Text (text).
 * - wether they are inherited by this node's children (flow) or not (retain).
 *
 * @public
 */
export interface TStylesShape {
  readonly nativeTextFlow: CSSProcessedProps['native']['text']['flow'];
  readonly nativeBlockFlow: CSSProcessedProps['native']['block']['flow'];
  readonly nativeTextRet: CSSProcessedProps['native']['text']['retain'];
  readonly nativeBlockRet: CSSProcessedProps['native']['block']['retain'];
  readonly webTextFlow: CSSProcessedProps['web']['text']['flow'];
  readonly webBlockRet: CSSProcessedProps['web']['block']['retain'];
}

export class TStyles implements TStylesShape {
  public readonly nativeTextFlow: CSSProcessedProps['native']['text']['flow'];
  public readonly nativeBlockFlow: CSSProcessedProps['native']['block']['flow'];
  public readonly nativeTextRet: CSSProcessedProps['native']['text']['retain'];
  public readonly nativeBlockRet: CSSProcessedProps['native']['block']['retain'];
  public readonly webTextFlow: CSSProcessedProps['web']['text']['flow'];
  public readonly webBlockRet: CSSProcessedProps['web']['block']['retain'];

  constructor(
    ownProcessedProps: CSSProcessedProps,
    parentStyles?: TStyles | null
  ) {
    this.nativeTextFlow = inheritProperties(
      parentStyles?.nativeTextFlow,
      ownProcessedProps.native.text.flow
    );
    this.nativeBlockFlow = inheritProperties(
      parentStyles?.nativeBlockFlow,
      ownProcessedProps.native.block.flow
    );
    this.webTextFlow = inheritProperties(
      parentStyles?.webTextFlow,
      ownProcessedProps.web.text.flow
    );
    // In theory, we shouldn't merge those properties. However, these
    // properties being textDecoration*, we actually want children nodes to
    // inherit from them. A cleaner solution would be to to let each TNode
    // handle its merging logic, because only TPhrasing and TText nodes would
    // need to merge these.
    this.nativeTextRet = inheritProperties(
      parentStyles?.nativeTextRet,
      ownProcessedProps.native.text.retain
    );
    this.nativeBlockRet = ownProcessedProps.native.block.retain;
    this.webBlockRet = ownProcessedProps.web.block.retain;
  }

  static empty(): TStyles {
    return new TStyles(new CSSProcessedProps());
  }
}
