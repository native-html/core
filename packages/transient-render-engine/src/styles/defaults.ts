import { StylesConfig } from './types';

export const defaultStylesConfig: Required<StylesConfig> = {
  classesStyles: {},
  enableCSSInlineProcessing: true,
  enableUserAgentStyles: false,
  idsStyles: {},
  tagsStyles: {},
  baseStyle: {
    fontSize: 14,
    whiteSpace: 'normal'
  }
};
