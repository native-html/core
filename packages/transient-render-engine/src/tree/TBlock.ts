import { TNode, TNodeInit } from './TNode';
import { SerializableElement } from '../dom/to-serializable';
import HTMLContentModel from '../model/HTMLContentModel';

export interface TBlockInit extends TNodeInit {}

export class TBlock extends TNode {
  public domNode?: SerializableElement;
  public readonly displayName: string = 'TBlock';
  constructor(init: TBlockInit) {
    super(init, 'block');
    if (init.domNode) {
      this.domNode = init.domNode;
    }
  }

  matchContentModel(contentModel: HTMLContentModel) {
    return (
      contentModel === HTMLContentModel.block ||
      contentModel === HTMLContentModel.mixed
    );
  }
}
