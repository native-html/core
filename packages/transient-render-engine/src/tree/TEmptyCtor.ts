import HTMLContentModel from '../model/HTMLContentModel';
import { Element } from '../dom/dom-utils';
import { TNodeImpl, TNodeInit } from './tree-types';
import TNodeCtor, { GenericTNodeCtor, Mutable } from './TNodeCtor';

export interface TEmptyImpl extends TNodeImpl<TEmptyInit> {
  readonly domNode: Element;
}

export interface TEmptyInit extends TNodeInit {
  domNode: Element;
}

const TEmptyCtor = function TEmpty(
  this: Mutable<TEmptyImpl>,
  init: TEmptyInit
) {
  this.initialize(init);
} as Function as GenericTNodeCtor<TEmptyInit, TEmptyImpl>;

//@ts-ignore
TEmptyCtor.prototype = new TNodeCtor('empty', 'TEmpty');

TEmptyCtor.prototype.matchContentModel = function matchContentModel(
  contentModel: HTMLContentModel
) {
  return contentModel === HTMLContentModel.none;
};

export default TEmptyCtor;

export { TEmptyCtor };
