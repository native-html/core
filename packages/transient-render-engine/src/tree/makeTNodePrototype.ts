import { TNodeImpl, TNodeInit, TNodeInvariants, TNodeType } from './tree-types';

export type TNodeCtor<Init = TNodeInit, Impl = TNodeImpl> = {
  new (init: Init): Impl;
  prototype: Impl;
};

export type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

function updateNodeIndexes(node: Mutable<TNodeImpl>, i: number) {
  node.nodeIndex = i;
}

const emptyAttrs = {};

const prototype: Omit<TNodeImpl, keyof TNodeInvariants> & {
  // TODO remove
  className: string;
} = {
  children: Object.freeze([]) as any,
  init: Object.freeze({}) as any,
  __classes: null,
  __styles: null,
  __nodeIndex: null,
  __trimmedLeft: false,
  __trimmedRight: false,
  get attributes() {
    return this.domNode?.attribs || emptyAttrs;
  },

  get hasWhiteSpaceCollapsingEnabled() {
    return typeof this.styles.webTextFlow.whiteSpace === 'string'
      ? this.styles.webTextFlow.whiteSpace === 'normal'
      : true;
  },

  get contentModel() {
    return this.elementModel?.contentModel || null;
  },

  get parentStyles() {
    return this.init.parentStyles || this.parent?.styles || null;
  },

  get id() {
    return this.attributes.id || null;
  },

  get classes() {
    if (this.__classes === null) {
      this.__classes = this.attributes.class?.split(/\s+/) || [];
    }
    return this.__classes;
  },

  get className() {
    // TODO remove
    return this.attributes.class;
  },

  get domNode() {
    return this.init.domNode || null;
  },

  get elementModel() {
    return this.init.elementModel;
  },

  get stylesMerger() {
    return this.init.stylesMerger;
  },

  get styles() {
    if (this.__styles === null) {
      const self = (this as unknown) as TNodeImpl;
      this.__styles =
        self.init.styles ||
        self.init.stylesMerger.buildStyles(
          self.attributes.style,
          self.parentStyles || null,
          (this as unknown) as any
        );
    }
    return this.__styles;
  },

  get tagName() {
    return this.init.domNode?.name || null;
  },

  get parent() {
    return this.init.parent || null;
  },

  get nodeIndex() {
    if (this.__nodeIndex === null) {
      this.__nodeIndex = this.init.nodeIndex || 0;
    }
    return this.__nodeIndex;
  },

  set nodeIndex(nodeIndex: number) {
    this.__nodeIndex = nodeIndex;
  },

  bindChildren(children, shouldUpdateNodeIndexes = false) {
    //@ts-ignore
    this.children = children;
    if (shouldUpdateNodeIndexes) {
      children.forEach(updateNodeIndexes);
    }
  },

  cloneInitParams(partial) {
    return Object.assign({}, this.init, partial) as any;
  },

  isCollapsibleLeft() {
    if (this.children.length) {
      return (
        this.hasWhiteSpaceCollapsingEnabled &&
        this.children[0].isCollapsibleLeft()
      );
    }
    return false;
  },

  isCollapsibleRight() {
    if (this.children.length) {
      return (
        this.hasWhiteSpaceCollapsingEnabled &&
        this.children[this.children.length - 1].isCollapsibleRight()
      );
    }
    return false;
  },

  isWhitespace() {
    return false;
  },

  isEmpty() {
    return false;
  },

  trimLeft() {
    if (!this.__trimmedLeft && this.children.length) {
      const firstChild = this.children[0];
      firstChild.trimLeft();
      if (firstChild.isEmpty()) {
        //@ts-ignore
        this.children.splice(0, 1);
      }
      this.__trimmedLeft = true;
    }
  },

  trimRight() {
    if (!this.__trimmedRight && this.children.length) {
      const lastChild = this.children[this.children.length - 1];
      lastChild.trimRight();
      if (lastChild.isEmpty()) {
        //@ts-ignore
        this.children.splice(-1, 1);
      }
      this.__trimmedRight = true;
    }
  },

  matchContentModel() {
    return false;
  },

  spliceChildren(indexesToSplice) {
    if (indexesToSplice) {
      let offset = 0;
      for (const i of indexesToSplice) {
        //@ts-ignore
        this.children.splice(i - offset, 1);
        offset += 1;
      }
    }
  },

  collapse(params) {
    this.collapseChildren(params);
    this.bindChildren(this.children, true);
  },

  collapseChildren() {
    return;
  }
};

export function initialize<Impl extends TNodeImpl<any> = TNodeImpl>(
  self: Mutable<Impl>,
  init: Impl['init']
) {
  // self.init = init;
  Object.defineProperty(self, 'init', {
    get: () => init,
    enumerable: false
  });
}

const PCtor = (function PCtor(
  this: Mutable<TNodeImpl>,
  type: TNodeType,
  displayName: string
) {
  this.type = type;
  this.displayName = displayName;
} as unknown) as {
  new (type: TNodeType, displayName: string): TNodeImpl;
  prototype: TNodeImpl;
};

PCtor.prototype = (prototype as unknown) as TNodeImpl;

export default function makeTNodePrototype<Impl = TNodeImpl>(
  type: TNodeType,
  displayName: string,
  extraAccessors?: {
    [k in Exclude<keyof Impl, keyof TNodeImpl>]?: {
      get: () => any;
      set?: (val: any) => void;
    };
  }
): Impl {
  //@ts-ignore
  return Object.create(new PCtor(type, displayName), extraAccessors);
}
