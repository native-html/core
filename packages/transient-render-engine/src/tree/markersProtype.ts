import { Markers } from './tree-types';

const markersProtype: Markers = {
  anchor: false,
  edits: 'none',
  lang: 'en',
  olNestLevel: -1,
  ulNestLevel: -1,
  direction: 'ltr',
  extend(this: Markers) {
    return Object.create(this);
  },
  toString() {
    let print = 'Markers {\n';
    for (const key in this) {
      //@ts-ignore
      const val = this[key];
      if (typeof val !== 'function') {
        print += `  ${key}: ${JSON.stringify(val)}\n`;
      }
    }
    return print + '}';
  }
};

export default markersProtype;
