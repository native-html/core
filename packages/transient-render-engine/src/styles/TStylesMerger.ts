import CSSProcessor, {
  CSSProcessedProps,
  CSSProcessorConfig,
  MixedStyleDeclaration
} from '@native-html/css-processor';
import HTMLModelRegistry from '../model/HTMLModelRegistry';
import { Markers } from '../tree/tree-types';
import { TStyles } from './TStyles';
import { StylesConfig } from './types';

export const emptyProcessedPropsReg: CSSProcessedProps =
  new CSSProcessedProps();

function mapMixedStyleRecordToCSSProcessedPropsReg(
  processor: CSSProcessor,
  styles?: Record<string, MixedStyleDeclaration>
): Record<string, CSSProcessedProps> {
  let regStyles: Record<string, CSSProcessedProps> = {};
  for (const key in styles) {
    regStyles[key] = processor.compileStyleDeclaration(styles[key]);
  }
  return regStyles;
}

// Specificity hierarchy, in descending order:
// 1. Inline styles
// 2. ID (idsStyles)
// 3. Classes (classesStyles)
// 4. Element name (tagsStyles)
// 5. Attribute styles (styles derived from attributes)
// 6. Default element styles
// 7. Inherited styles (baseFontStyles)
export class TStylesMerger {
  private processor: CSSProcessor;
  private tagsStyles: Record<string, CSSProcessedProps>;
  private classesStyles: Record<string, CSSProcessedProps>;
  private idsStyles: Record<string, CSSProcessedProps>;
  private enableCSSInlineProcessing: boolean;
  private enableUserAgentStyles: boolean;
  private modelRegistry: HTMLModelRegistry<string>;
  constructor(
    config: Required<StylesConfig>,
    modelRegistry: HTMLModelRegistry<string>,
    cssProcessorConfig?: CSSProcessorConfig
  ) {
    this.processor = new CSSProcessor(cssProcessorConfig);
    this.modelRegistry = modelRegistry;
    this.classesStyles = mapMixedStyleRecordToCSSProcessedPropsReg(
      this.processor,
      config.classesStyles
    );
    this.tagsStyles = mapMixedStyleRecordToCSSProcessedPropsReg(
      this.processor,
      config.tagsStyles
    );
    this.idsStyles = mapMixedStyleRecordToCSSProcessedPropsReg(
      this.processor,
      config.idsStyles
    );
    this.enableCSSInlineProcessing = config.enableCSSInlineProcessing;
    this.enableUserAgentStyles = config.enableUserAgentStyles;
  }

  compileInlineCSS(inlineCSS: string) {
    return this.processor.compileInlineCSS(inlineCSS);
  }

  compileStyleDeclaration(styleDeclaration: MixedStyleDeclaration) {
    return this.processor.compileStyleDeclaration(styleDeclaration);
  }

  buildStyles(
    inlineStyle: string,
    parentStyles: TStyles | null,
    descriptor: {
      tagName: string | null;
      classes: string[];
      id: string | null;
      attributes: Record<string, string>;
      markers: Markers;
    }
  ): TStyles {
    const ownInlinePropsReg =
      this.enableCSSInlineProcessing && inlineStyle
        ? this.compileInlineCSS(inlineStyle)
        : null;
    const model = descriptor.tagName
      ? this.modelRegistry.getElementModelFromTagName(descriptor.tagName)
      : null;
    const userTagOwnProps =
      this.tagsStyles[descriptor.tagName as string] ?? null;
    const userIdOwnProps = this.idsStyles[descriptor.id as string] ?? null;
    const classes = descriptor.classes;
    const userClassesOwnPropsList = classes.map(
      (c) => this.classesStyles[c] || null
    );
    const derivedPropsFromAttributes = this.enableUserAgentStyles
      ? model?.getUADerivedCSSProcessedPropsFromAttributes(
          descriptor.attributes,
          descriptor.markers
        ) || null
      : null;
    const userAgentTagProps = this.enableUserAgentStyles
      ? model?.mixedUAStyles ?? null
      : null;
    // Latest properties will override former properties.
    const mergedOwnProps = emptyProcessedPropsReg.merge(
      userAgentTagProps &&
        this.processor.compileStyleDeclaration(userAgentTagProps),
      derivedPropsFromAttributes &&
        this.processor.compileStyleDeclaration(derivedPropsFromAttributes),
      userTagOwnProps,
      ...userClassesOwnPropsList,
      userIdOwnProps,
      ownInlinePropsReg
    );
    return new TStyles(mergedOwnProps, parentStyles);
  }
}
