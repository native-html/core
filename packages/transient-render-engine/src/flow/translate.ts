import { TNodeImpl, TNodeInit } from '../tree/tree-types';
import { Document } from 'domhandler';
import { TTextCtor } from '../tree/TTextCtor';
import { TPhrasingCtor } from '../tree/TPhrasingCtor';
import { TBlockCtor } from '../tree/TBlockCtor';
import { TDocumentCtor, TDocumentImpl } from '../tree/TDocumentImpl';
import { isDomElement, isDomText, Element, Node, Text } from '../dom/dom-utils';
import { TStyles } from '../styles/TStyles';

import { TEmptyCtor } from '../tree/TEmptyCtor';
import { DataFlowParams } from './types';

export function mapNodeList({
  nodeList,
  parent,
  parentStyles,
  params
}: {
  nodeList: Node[];
  parentStyles?: TStyles;
  parent: TNodeImpl | null;
  params: DataFlowParams;
}): TNodeImpl[] {
  const nextMap: TNodeImpl[] = [];
  for (const i in nodeList) {
    const child = nodeList[i];
    const translated = translateNode({
      node: child,
      parentStyles,
      params,
      nodeIndex: Number(i),
      parent
    });
    if (translated) {
      nextMap.push(translated);
    }
  }
  return nextMap;
}

export function bindChildren(
  node: TNodeImpl,
  children: Node[],
  params: DataFlowParams
) {
  if (!node.elementModel || !node.elementModel.isOpaque) {
    node.bindChildren(
      mapNodeList({
        nodeList: children,
        parent: node,
        params
      })
    );
  }
}

function translateElement({
  node,
  nodeIndex,
  params,
  parent,
  parentStyles
}: TranslateParams<Element>): TNodeImpl | null {
  const tagName = node.tagName;
  const sharedProps: Omit<TNodeInit, 'contentModel' | 'elementModel'> = {
    nodeIndex,
    parentStyles,
    context: params,
    domNode: node,
    parent
  };
  const elementModel = params.modelRegistry.getElementModelFromTagName(tagName);
  if (!elementModel) {
    return new TEmptyCtor({
      ...sharedProps,
      isUnregistered: true,
      elementModel,
      domNode: node
    });
  }
  if (elementModel.isTranslatableTextual()) {
    if (node.children.length === 1) {
      const child = node.children[0] as Node;
      if (isDomText(child)) {
        return new TTextCtor({
          ...sharedProps,
          elementModel,
          textNode: child,
          domNode: node
        });
      }
    } else if (node.children.length === 0) {
      return new TTextCtor({
        ...sharedProps,
        elementModel,
        domNode: node,
        textNode: new Text('')
      });
    }
    const phrasing = new TPhrasingCtor({
      ...sharedProps,
      domNode: node,
      elementModel
    });
    bindChildren(phrasing, node.children, params);
    return phrasing;
  }
  if (elementModel.isTranslatableBlock()) {
    const block = new TBlockCtor({
      ...sharedProps,
      elementModel,
      parentStyles,
      domNode: node
    });
    bindChildren(block, node.children, params);
    return block;
  }
  return new TEmptyCtor({
    ...sharedProps,
    isUnregistered: false,
    elementModel,
    domNode: node
  });
}

interface TranslateParams<T = Node> {
  node: T;
  params: DataFlowParams;
  parent: TNodeImpl | null;
  parentStyles?: TStyles;
  nodeIndex?: number;
}

export function translateNode({
  node,
  parentStyles,
  params,
  nodeIndex,
  parent
}: TranslateParams<Node | null>): TNodeImpl | null {
  if (isDomText(node)) {
    return new TTextCtor({
      textNode: node,
      context: params,
      parentStyles,
      domNode: null,
      elementModel: null,
      nodeIndex,
      parent
    });
  }
  if (isDomElement(node)) {
    return translateElement({
      node,
      parentStyles,
      params,
      nodeIndex,
      parent
    });
  }
  return null;
}

export function translateDocument(
  document: Document,
  params: DataFlowParams
): TDocumentImpl {
  const tdoc = new TDocumentCtor({
    context: params,
    styles: params.baseStyles,
    domNode: document as any
  });
  bindChildren(tdoc, document.children, params);
  tdoc.parseChildren();
  return tdoc;
}
