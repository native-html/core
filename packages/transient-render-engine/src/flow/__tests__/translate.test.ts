import { TStyles } from '../../styles/TStyles';
import { TEmptyCtor } from '../../tree/TEmptyCtor';
import { mapNodeList } from '../translate';
import { rfc002Source } from './shared';
import {
  defaultModelRegistry,
  defaultStylesMerger,
  translateTreeTest
} from './utils';

describe('translateNode function', () => {
  it('should comply with RFC002 example (translating)', () => {
    const ttree = translateTreeTest(rfc002Source);
    expect(ttree).toMatchSnapshot();
  });
  it('should translate styles (1)', () => {
    const ttree = translateTreeTest('<div style="font-size: 18px"></div>');
    expect(ttree).toMatchSnapshot();
  });
  it('should translate styles (2)', () => {
    const tdoc = translateTreeTest(
      '<a href="http://google.fr"><div style="background-color: red;"></div></a>'
    );
    expect(tdoc).toMatchSnapshot();
  });
  it('should pass regression with rgb() including spaces', () => {
    const span = translateTreeTest(
      '<span style="color: rgb(101, 123, 131);">A</span>'
    );
    expect(span).toMatchSnapshot();
  });
  it('should set parent', () => {
    const tdoc = translateTreeTest(rfc002Source);
    expect(tdoc.children).toHaveLength(7);
    for (const child of tdoc.children) {
      expect(child.parent).toBe(tdoc);
    }
  });
  it('should translate a phrasing element with one text child node to a TText element', () => {
    const ttree = translateTreeTest('<span><strong>Hello!</strong></span>');
    expect(ttree).toMatchSnapshot();
  });
  it('should ignore unregistered tags', () => {
    const ttree = translateTreeTest('<nonexistingtag></nonexistingtag>');
    expect(ttree).toMatchSnapshot();
  });
  describe('regarding opaque blocks', () => {
    const svgSrc = '<svg><path /></svg>';
    const ttree = translateTreeTest(svgSrc);
    it('should translate to TEmpty', () => {
      expect(ttree).toBeInstanceOf(TEmptyCtor);
    });
    it('should set domNode attribute', () => {
      expect(ttree.domNode).toBeDefined();
    });
    it('should not translate children', () => {
      expect(ttree.children).toHaveLength(0);
    });
  });
  describe('should add a `nodeIndex` field with the index of this node relative to its parent', () => {
    const src = `<table>
    <tr>
      <th>Month</th>
      <th>Fortnight</th>
      <th>Savings</th>
    </tr>
    <tr>
      <td>January</td>
      <td>first</td>
      <td >$50</td>
    </tr>
    <tr>
      <td >second</td>
      <td>$80</td>
      <td>$80</td>
    </tr>
  </table>`;
    const ttree = translateTreeTest(src);
    expect(ttree.tagName).toBe('table');
    ttree.children.forEach((child, i) => {
      expect(child.nodeIndex).toBe(i);
    });
  });
  describe('regarding markers', () => {
    it('should set an anchor marker with <a> elements', () => {
      const ttree = translateTreeTest('<a href="hello">Yo!</a>');
      expect(ttree.markers.anchor).toBe(true);
      ttree.children.forEach((c) => {
        expect(c.markers.anchor).toBe(true);
      });
    });
    it('should set an ulNestLevel marker with <ul> elements', () => {
      const ttree = translateTreeTest('<ul><li></li></ul>');
      expect(ttree.markers.ulNestLevel).toBe(0);
      ttree.children.forEach((c) => {
        expect(c.markers.ulNestLevel).toBe(0);
      });
    });
    it('should set an olNestLevel marker with <ol> elements', () => {
      const ttree = translateTreeTest('<ol><li></li></ol>');
      expect(ttree.markers.olNestLevel).toBe(0);
      ttree.children.forEach((c) => {
        expect(c.markers.olNestLevel).toBe(0);
      });
    });
    it('should set an edit marker with <ins> elements', () => {
      const ttree = translateTreeTest('<ins>Text</ins>');
      expect(ttree.markers.edits).toBe('ins');
      ttree.children.forEach((c) => {
        expect(c.markers.edits).toBe('ins');
      });
    });
    it('should set an edit marker with <del> elements', () => {
      const ttree = translateTreeTest('<del>Text</del>');
      expect(ttree.markers.edits).toBe('del');
      ttree.children.forEach((c) => {
        expect(c.markers.edits).toBe('del');
      });
    });
    it('should set a lang marker with lang attributes', () => {
      const ttree = translateTreeTest('<div lang="fr">Text</div>');
      expect(ttree.markers.lang).toBe('fr');
      ttree.children.forEach((c) => {
        expect(c.markers.lang).toBe('fr');
      });
    });
    it('should set a direction marker with dir attributes', () => {
      const ttree = translateTreeTest('<div dir="rtl">Text</div>');
      expect(ttree.markers.direction).toBe('rtl');
      ttree.children.forEach((c) => {
        expect(c.markers.direction).toBe('rtl');
      });
    });
  });
});

describe('mapNodeList function', () => {
  it('should ignore untranslated nodes', () => {
    expect(
      mapNodeList({
        nodeList: [null] as any,
        parent: null,
        params: {
          stylesMerger: defaultStylesMerger,
          modelRegistry: defaultModelRegistry,
          baseStyles: TStyles.empty(),
          removeLineBreaksAroundEastAsianDiscardSet: false
        }
      })
    ).toHaveLength(0);
  });
});
