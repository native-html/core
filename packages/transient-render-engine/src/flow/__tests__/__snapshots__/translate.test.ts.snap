// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateNode function should comply with RFC002 example (translating) 1`] = `
<TPhrasing tagName="a" nodeIndex={0} href="https://domain.com">
  <TText anonymous nodeIndex={0} data="\\nThis is\\n" />
  <TText tagName="span" nodeIndex={1} data="phrasing content" />
  <TText anonymous nodeIndex={2} data="\\n" />
  <TBlock tagName="img" nodeIndex={3} src="https://domain.com/logo.jpg" />
  <TText anonymous nodeIndex={4} data="\\n    and this is " />
  <TText tagName="strong" nodeIndex={5} data="too" />
  <TText anonymous nodeIndex={6} data=".\\n" />
</TPhrasing>
`;

exports[`translateNode function should ignore unregistered tags 1`] = `<TEmpty tagName="nonexistingtag" unregistered nodeIndex={0} />`;

exports[`translateNode function should pass regression with rgb() including spaces 1`] = `<TText tagName="span" nodeIndex={0} data="A" nativeStyles={{ color: "rgb(101, 123, 131)" }} />`;

exports[`translateNode function should translate a phrasing element with one text child node to a TText element 1`] = `
<TPhrasing tagName="span" nodeIndex={0}>
  <TText tagName="strong" nodeIndex={0} data="Hello!" />
</TPhrasing>
`;

exports[`translateNode function should translate styles (1) 1`] = `<TBlock tagName="div" nodeIndex={0} />`;

exports[`translateNode function should translate styles (2) 1`] = `
<TPhrasing tagName="a" nodeIndex={0} href="http://google.fr">
  <TBlock tagName="div" nodeIndex={0} nativeStyles={{ backgroundColor: "red" }} />
</TPhrasing>
`;
